File: Haskell/PPM.hs
{- 
  For GRHUG
  Version: July 2009
  Creative Commons Attribution-Share Alike 3.0 United States License
-}

module PPM (ppmPrefix, 
            blackOnWhite, whiteOnBlack, 
            grayScale, redScale, greenScale, blueScale,
            randomColors, randomColorsGenerator)
where
  
import System.Random
import Data.List
import Complex

import Fractals

max_color = 255
  
ppmPrefix (Dimension width height) =
  ["P3", show width, show height, show max_color]

ppmEntry :: Integer -> Integer -> Integer -> String
ppmEntry r g b = (show r) ++ " " ++ (show g) ++ " " ++ (show b)
ppmEntryGray n = ppmEntry n n n

white = ppmEntryGray max_color
black = ppmEntryGray 0

-----------------------------------------------------------
-- helpers
-----------------------------------------------------------
smoother z n = fromInteger n + 1.0 - log (log  (magnitude z)) / log 2.0


-----------------------------------------------------------
-- black, white, and gray
-----------------------------------------------------------
blackOnWhite Inside = black
blackOnWhite (Outside _ _) = white

whiteOnBlack Inside = white
whiteOnBlack (Outside _ _) = black

grayScale Inside = black
grayScale (Outside z n) = ppmEntryGray val
  where
    val = round $ fromInteger max_color * (sqrt $ fromInteger n / fromInteger max_iters)^2


-----------------------------------------------------------
-- basic primary-color scales
-- coloring scheme from http://warp.povusers.org/Mandelbrot/
-----------------------------------------------------------
redScale   = povScale (\c -> ppmEntry c 0 0) (\c -> ppmEntry max_color c c)
greenScale = povScale (\c -> ppmEntry 0 c 0) (\c -> ppmEntry c max_color c)
blueScale  = povScale (\c -> ppmEntry 0 0 c) (\c -> ppmEntry c c max_color)

povScale _ _ Inside = black
povScale plateau border (Outside _ n)
  | n < half_iters = plateau (scaleIter n)
  | otherwise      = border $ scaleIter $ n - half_iters
  where
    half_iters = (div max_iters 2) - 1
    scaleIter i = round $ iterRatio i * fromInteger max_color
    iterRatio i = fromInteger (2*(i-1)) / fromInteger max_iters


-----------------------------------------------------------
-- random colors
-----------------------------------------------------------
randomColors _      Inside        = black  
randomColors colors (Outside _ n) = genericIndex colors n
    
randomColorsGenerator seed = colorStream $ randomRs (0, max_color) (mkStdGen seed)
  where
    colorStream (r : g : b : rest) = ppmEntry r g b : colorStream rest

File: Erlang/ppm.erl
-module(ppm).
-export([black_on_white/1, white_on_black/1, gray_scale/1, red_scale/1, green_scale/1, blue_scale/1, random_colors/0, write_ppm/3]).

-include("fractal_constants.hrl").

black_on_white({inside})        -> black();
black_on_white({outside, _, _}) -> white().

white_on_black({inside})        -> white();
white_on_black({outside, _, _}) -> black().

gray_scale({inside}) -> black();
gray_scale({outside, _, N}) ->
     Val = erlang:round(?MAX_COLOR * (N / ?MAX_ITERS)),
     ppm_entry_gray(Val).

red_scale(X)   -> pov_scale(fun (C) -> ppm_entry(C, 0, 0) end, fun (C) -> ppm_entry(?MAX_COLOR, C, C) end, X).
green_scale(X) -> pov_scale(fun (C) -> ppm_entry(0, C, 0) end, fun (C) -> ppm_entry(C, ?MAX_COLOR, C) end, X).
blue_scale(X)  -> pov_scale(fun (C) -> ppm_entry(0, 0, C) end, fun (C) -> ppm_entry(C, C, ?MAX_COLOR) end, X).

random_colors() ->
    random:seed(),
    Colors = random_colors_array(?MAX_ITERS),
    fun (X) ->
        case X of
            {inside}        -> black();
            {outside, _, N} -> lists:nth((N rem length(Colors)) + 1, Colors)
        end
    end.


%% **********
%% Helpers...

random_colors_array(0) -> [];
random_colors_array(N) ->
    Color = ppm_entry(random:uniform(?MAX_COLOR-1), random:uniform(?MAX_COLOR-1), random:uniform(?MAX_COLOR-1)),
    [Color | random_colors_array(N-1)].

white() -> ppm_entry_gray(?MAX_COLOR).
black() -> ppm_entry_gray(0).

ppm_entry(R, G, B) -> string:join([integer_to_list(R), integer_to_list(G), integer_to_list(B)], " ").
ppm_entry_gray(N) -> ppm_entry(N, N, N).

pov_scale(_, _, {inside}) -> black();
pov_scale(Plateau, Border, {outside, _, N}) ->
    HalfIters = ?MAX_ITERS / 2,
    if
        N < HalfIters -> Plateau(scaleIter(N));
        true          -> Border(scaleIter(N - HalfIters))
    end.

scaleIter(I) -> erlang:round(iterRatio(I) * ?MAX_COLOR).
iterRatio(I) -> (2 * (I-1)) / ?MAX_ITERS.

write_ppm(Filename, Colors, [Width, Height]) ->
    {ok, S} = file:open(Filename, write),
    io:format(S, "P3~n", []),
    io:format(S, "~p ~p~n", [Width, Height]),
    io:format(S, "~p~n", [?MAX_COLOR]),
    lists:foreach(
        fun(Row) ->
            lists:foreach(fun(C) -> io:format(S, "~s ", [C]) end, Row),
            io:format(S, "~n", [])
        end,
        Colors),
    file:close(S).

